.github/workflows/gpu-ci.yml

```
@@ -0,0 +1,78 @@
+name: GPU jobs
+
+on:
+  push:
+    branches:
+      - maintenance/**
+  pull_request:
+    branches:
+      - main
+      - maintenance/**
+
+permissions:
+   contents: read  # to fetch code (actions/checkout)
+
+env:
+  CCACHE_DIR: "${{ github.workspace }}/.ccache"
+  PIXI_CACHE_DIR: "${{ github.workspace }}/.cache/rattler"
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+
+jobs:
+  get_commit_message:
+    name: Get commit message
+    uses: ./.github/workflows/commit_message.yml
+    if: >
+      needs.get_commit_message.outputs.message == 1
+      && (github.repository == 'scipy/scipy' || github.repository == '')
+
+  pytorch_gpu:
+    name: PyTorch, JAX, CuPy GPU
+    needs: get_commit_message
+    runs-on: ghcr.io/cirruslabs/ubuntu-runner-amd64-gpu:22.04
+    steps:
+      - name: Checkout scipy repo
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+        with:
+          submodules: recursive
+
+      - name: Cache pixi
+        uses: cirruslabs/cache@v4  #caa3ad0624c6c2acd8ba50ad452d1f44bba078bb # v4
+        with:
+          path: ${{ env.PIXI_CACHE_DIR }}
+          key: ${{ runner.os }}-gpu-pixi
+
+      - name: Setup compiler cache
+        uses: cirruslabs/cache@v4  #caa3ad0624c6c2acd8ba50ad452d1f44bba078bb # v4
+        with:
+          path: ${{ env.CCACHE_DIR }}
+          key: ${{ runner.os }}-gpu-ccache
+
+      - name: run nvidia-smi
+        run: nvidia-smi
+
+      - name: run nvidia-smi --query
+        run: nvidia-smi --query
+
+      - uses: prefix-dev/setup-pixi@ba3bb36eb2066252b2363392b7739741bb777659 # v0.8.1
+        with:
+          pixi-version: v0.39.2
+          manifest-path: .github/workflows/pixi.toml
```

It sounds like this is needed to ensure caching of the non-default environments: https://github.com/prefix-dev/setup-pixi?tab=readme-ov-file#install-multiple-environments-in-one-job

```suggestion
      - uses: prefix-dev/setup-pixi@ba3bb36eb2066252b2363392b7739741bb777659 # v0.8.1
        with:
          pixi-version: v0.39.2
          manifest-path: .github/workflows/pixi.toml
          environments: >-
            default
            torch-cuda
            jax-cuda
            cupy
```

---

unless this was deliberate to avoid hitting the cache limit?

---

It's caching everything it looks like - there's 3 GB worth of packages cached says the CI log, and there's no downloads on using the other envs. 

This doesn't quite work the way you think - `setup-pixi` does no caching, it's done higher up by `cirruslabs/cache@v4`. Pixi already has a local cache of all packages that were downloaded in a previous successful run, so no envs need specifying.

---

Ah okay, so setup-pixi *can* do caching, but we're using cirruslabs/cache instead, so it would be redundant - is that right?

---

Yes, that is correct. The README at https://github.com/cirruslabs/cache explains the benefit over regular caching (tl;dr it's more local and hence faster, no need to cycle between Cirrus and GitHub infra)

---

should we use `cache: false` with setup-pixi, then?

---

Yes, good point - adding that.